cobalt/build/build_info.py

```diff
@@ -0,0 +1,120 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Generates a Cobalt Build Info json."""
+
+import datetime
+import json
+import os
+import re
+import subprocess
+import sys
+
+_FILE_DIR = os.path.dirname(__file__)
+_COMMIT_COUNT_BUILD_ID_OFFSET = 1000000
+
+# git log --grep can't handle complex patterns.
+_BUILD_ID = r'[1-9]' + r'[0-9]' * 6 + r'[0-9]*'
+_GIT_REV = r'[0-9a-f]' * 40
+
+# git log --grep can't handle capture groups.
+_BUILD_ID_WITH_CAPTURE = f'({_BUILD_ID})'
+_GIT_REV_WITH_CAPTURE = f'({_GIT_REV})'
+
+_BUILD_ID_PATTERN = r'^BUILD_NUMBER={}$'
+_GIT_REV_PATTERN = r'^GitOrigin-RevId: {}$'
+
+
+def get_build_id_and_git_rev_from_commits(cwd):
+  # Build id and git rev must come from the same commit.
+  search_pattern = _BUILD_ID_PATTERN.format(_BUILD_ID)
+  output = subprocess.check_output(
+      ['git', 'log', '--grep', search_pattern, '-1', '--pretty=%b'],
+      cwd=cwd).decode()
+
+  # Gets build id.
+  compiled_build_id_pattern = re.compile(
+      _BUILD_ID_PATTERN.format(_BUILD_ID_WITH_CAPTURE), flags=re.MULTILINE)
+  match_build_id = compiled_build_id_pattern.search(output)
+
+  if match_build_id is None:
+    return None
+  build_id = match_build_id.group(1)
+
+  # Gets git rev.
+  compiled_git_rev_pattern = re.compile(
+      _GIT_REV_PATTERN.format(_GIT_REV_WITH_CAPTURE), flags=re.MULTILINE)
+  match_git_rev = compiled_git_rev_pattern.search(output)
+
+  if match_git_rev is None:
+    git_rev = ''
+  else:
+    git_rev = match_git_rev.group(1)
+
+  return (build_id, git_rev)
+
+
+def get_build_id_from_commit_count(cwd):
+  output = subprocess.check_output(['git', 'rev-list', '--count', 'HEAD'],
+                                   cwd=cwd)
+  build_id = int(output.strip().decode()) + _COMMIT_COUNT_BUILD_ID_OFFSET
+  return build_id
+
+
+def get_from_last_commit(placeholder, cwd):
+  output = subprocess.check_output(
+      ['git', 'log', '-1', f'--pretty=format:{placeholder}'], cwd=cwd)
+  return output.strip().decode('utf-8')
+
+
+def get_hash_from_last_commit(cwd):
+  return get_from_last_commit(r'%H', cwd=cwd)
+
+
+def get_author_from_last_commit(cwd):
+  return get_from_last_commit(r'%an', cwd=cwd)
+
+
+def get_subject_from_last_commit(cwd):
+  return get_from_last_commit(r'%s', cwd=cwd)
+
+
+def build_info(output_path, cwd=_FILE_DIR):
+  """Writes a Cobalt build_info json file."""
+  build_rev = get_hash_from_last_commit(cwd=cwd)
+  build_id_git_rev = get_build_id_and_git_rev_from_commits(cwd=cwd)
```

Comment:
If you change the return value above this could be rewritten as 
`build_id, git_rev = get_build_id_and_git_rev_from_commits(cwd=cwd)`

You would have to refactor the code below slightly too.